pkg/ext/stores/tokens/tokens.go

```diff
@@ -0,0 +1,1376 @@
+	newToken.Status.TokenHash = ""
+	newToken.Status.TokenValue = tokenValue
+
+	return newToken, nil
+}
+
+func (t *Store) delete(ctx context.Context, token *ext.Token, options *metav1.DeleteOptions) error {
+	user, err := t.auth.UserName(ctx)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+	if !userMatch(user, token) {
+		return apierrors.NewNotFound(GVR.GroupResource(), token.Name)
+	}
+
+	return t.SystemStore.Delete(token.Name, options)
+}
+
+func (t *SystemStore) Delete(name string, options *metav1.DeleteOptions) error {
+	err := t.secretClient.Delete(TokenNamespace, name, options)
+	if err == nil {
+		return nil
+	}
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	return apierrors.NewInternalError(fmt.Errorf("failed to delete token %s: %w", name, err))
+}
+
+func (t *Store) get(ctx context.Context, name string, options *metav1.GetOptions) (*ext.Token, error) {
+	sessionID := t.auth.SessionID(ctx)
+
+	// note: have to get token first before we can check for a user mismatch
+	token, err := t.SystemStore.Get(name, sessionID, options)
+	if err != nil {
+		return nil, err
+	}
+
+	user, err := t.auth.UserName(ctx)
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+	if !userMatch(user, token) {
+		return nil, apierrors.NewNotFound(GVR.GroupResource(), name)
+	}
+
+	return token, nil
+}
+
+func (t *SystemStore) Get(name, sessionID string, options *metav1.GetOptions) (*ext.Token, error) {
+	// Core token retrieval from backing secrets
+	// We try to go through the fast cache as much as we can.
+
+	var err error
+	var currentSecret *corev1.Secret
+	empty := metav1.GetOptions{}
+	if options == nil || *options == empty {
+		currentSecret, err = t.secretCache.Get(TokenNamespace, name)
+	} else {
+		currentSecret, err = t.secretClient.Get(TokenNamespace, name, *options)
+	}
+
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to retrieve token %s: %w", name, err))
+	}
+	token, err := tokenFromSecret(currentSecret)
+	if err != nil {
+		return nil, apierrors.NewInternalError(fmt.Errorf("failed to extract token %s: %w", name, err))
+	}
+
+	token.Status.Current = token.Name == sessionID
+	token.Status.TokenValue = ""
+	return token, nil
+}
+
+func (t *Store) list(ctx context.Context, options *metav1.ListOptions) (*ext.TokenList, error) {
+	user, err := t.auth.UserName(ctx)
+	if err != nil {
+		return nil, apierrors.NewInternalError(err)
+	}
+
+	sessionID := t.auth.SessionID(ctx)
+
+	return t.SystemStore.list(false, user, sessionID, options)
+}
+
+func (t *SystemStore) List(options *metav1.ListOptions) (*ext.TokenList, error) {
+	return t.list(true, "", "", options)
+}
+
+func (t *SystemStore) list(fullView bool, user, sessionID string, options *metav1.ListOptions) (*ext.TokenList, error) {
+	// Core token listing from backing secrets
+	secrets, err := t.secretClient.List(TokenNamespace, *options)
```

Comment:
can we use the cache here if no options are provided? like we do for get

Reply:
maybe. will have to check the exact nature of the cache's list interface.

Reply:
Unfortunately not :disappointed: 

The regular client returns a `corev1.SecretList` structure containing `[]corev1Secret`.
The cache returns `[]*corev1.Secret`.

The slice difference does not matter. Just an additional loop to convert.

What matters is that our store has to return a `ListMeta.ResourceVersion` value in the token list it constructs.
When pulling from the direct client we can and do pass the `ResourceVersion` we get from the `SecretList`.
When reading from the cache however we simply do not have this data, breaking things.

Reply:
> When reading from the cache however we simply do not have this data, breaking things.


Why do we need to return `ListMeta.ResourceVersion` ? what breaks? We should figure out a way of using cache for listing. Not only for tokens, but for all resources using the imperative API

Reply:
Looking at https://kubernetes.io/docs/reference/using-api/api-concepts/ and searching for `ResourceVersion` it looks to be something to optimize watches, i.e. ask only for changes since the given resource version. I guess, although do not know, that leaving this out simply makes optimized requests impossible.

... Actually, reading deeper it may also break the ability for chunked retrieval of a large list via `kubectl`. Because the resource version may be encoded in the continue token used there.

Reply:
commits bb54d7267c70f2ed9c6c5ddedb31b506ac01789b, 057080b46ca21fb4b7cb096b8cc3c83bb8bf8962, and dc6ba5f9705a79704684bb1683ab153a81e65f03